src/path/mod.rs

```diff
@@ -1,33 +1,123 @@
 //! Path abstraction for Object Storage
 
 use itertools::Itertools;
+use snafu::{ensure, ResultExt, Snafu};
 use std::fmt::Formatter;
 
 /// The delimiter to separate object namespaces, creating a directory structure.
 pub const DELIMITER: &str = "/";
 
+/// The path delimiter as a single byte
+pub const DELIMITER_BYTE: u8 = DELIMITER.as_bytes()[0];
+
 mod parts;
 
 pub use parts::PathPart;
 
-/// An object storage location suitable for passing to cloud storage APIs such
-/// as AWS, GCS, and Azure.
+#[derive(Debug, Snafu)]
+#[allow(missing_docs)]
+pub enum ParseError {
+    #[snafu(display("Path \"{}\" contained empty path segment", path))]
+    EmptySegment { path: String },
+
+    #[snafu(display("Error parsing Path \"{}\": \"{}\"", path, source))]
+    BadSegment {
+        path: String,
+        source: parts::InvalidPart,
+    },
+}
+
+/// A parsed path representation that can be safely written to object storage
+///
+/// # Path Safety
+///
+/// In theory object stores support any UTF-8 character sequence, however, certain character
+/// sequences cause compatibility problems with some applications and protocols. As such the
+/// naming guidelines for [S3], [GCS] and [Azure Blob Storage] all recommend sticking to a
+/// limited character subset.
+///
+/// [S3]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+/// [GCS]: https://cloud.google.com/storage/docs/naming-objects
+/// [Azure Blob Storage]: https://docs.microsoft.com/en-us/rest/api/storageservices/Naming-and-Referencing-Containers--Blobs--and-Metadata#blob-names
+///
+/// This presents libraries with two options for consistent path handling:
+///
+/// 1. Allow constructing unsafe paths, allowing for both reading and writing of data to paths
+/// that may not be consistently understood or supported
+/// 2. Disallow constructing unsafe paths, ensuring data written can be consistently handled by
+/// all other systems, but preventing interaction with objects at unsafe paths
+///
+/// This library takes the second approach, in particular:
+///
+/// * Paths are delimited by `/`
+/// * Paths do not start with a `/`
```

Comment:
This might imply to a reader there there is no way to create a "absolute" path (e.g. an absolute path to a file). I don't think this is actually the case, but it might help to clarify

Reply:
This is actually the case, all paths are relative to the root of the object store - will clarify